### PR TITLE
Explore: Use SeriesData format for loki/logs

### LINKED ISSUE #16287: Use SeriesData format for loki/logging
We recently added SeriesData class that represents both TimeSeries and TableData equally well.  It would be great if the logs datasource used the same structure.  This would make it very easy for any datasource to implement the logs interface `SELECT time as ts, line FROM table` and would make loki usable in other visualisation panels without any other changes.

HUGE caveat, I have not used loki and don't know what I am talking about!  The demos look cool and I would love to be able to use explore logging with stuff that does not go through loki.

In #16285, I explored if SeriesData could transfer everything you need -- I think it can.  Check:
 * [logStreamToSeriesData](https://github.com/grafana/grafana/blob/ad1320b492038f82aed0eb803aef8c193e94ce6d/public/app/plugins/datasource/loki/result_transformer.ts#L112)
 * [seriesDataToLogStream](https://github.com/grafana/grafana/blob/ad1320b492038f82aed0eb803aef8c193e94ce6d/public/app/plugins/datasource/loki/result_transformer.ts#L126)

Ideally, loki could output standard SeriesData[] like:
```
{
    labels: {
      foo:'bar'
    },
    fields: [
      { name: 'ts', type: FieldType.time }, 
      { name: 'line', type: FieldType.string }
    ],
    rows: [
      ['1970-01-01T00:00:00Z','WARN somethign'],
      ['1970-01-01T00:00:00Z','INFO more'],
    ]
  };
```

Then the explore logs viewing component would take any SeriesData[] result (from loki/SQL/influx, whatever) and process it for use in the explore UI.  Maybe it converts to/from LogStream or maybe directly from SeriesData#row > LogRowModel.  Maybe we don't need LogRowModel, but just add more fields to SeriesData so you have everything you need

Essentially I am suggesting that a lot of what happens in loki datasource should happen in the view component, not the datasource.  Then we could use any datasource to load data without changing too much.

### PR DESCRIPTION
**Pull request by marefr**
**What this PR does / why we need it**:
This is the first step moving towards Explore supporting logs for additional datasources than Loki. In the first step we move all the log processing from Loki into Explore.

Summary of changes:
Make explore convert logs result returned from datasources to SeriesData, 
if needed, and for now convert SeriesData to LogsModel.
Loki datasource query now returns SeriesData and all
processing have been moved into explore instead.
Removed key from LogRowModel and use log row indexes as 
the unique key instead.
Removed id from LogsModel since it looks like it's not in use.

**Which issue(s) this PR fixes**:
Closes #16287

**Special notes for your reviewer**:
I removed `key` from `LogRowModel` and used index of each log row as key for rendering unique keys for `<LogRow />` component. To me this seems to work and makes the row model easier, but I may have missed some important aspect?

### CODE REVIEW

**Comment by marefr** [comment]
@ryantxu Please have a look at the changes if you're interested to see if this maps to your vision in referenced issue. This is the first step and there will probably be a lot of other changes needed, but now we make it easier for each datasource to return log data.

File: public/app/plugins/datasource/loki/datasource.ts
Comment (marefr) [inline public/app/plugins/datasource/loki/datasource.ts]: For now added these properties into meta field.

File: public/app/plugins/datasource/loki/result_transformer.ts
Comment (marefr) [inline public/app/plugins/datasource/loki/result_transformer.ts]: Decided to use time and message to make it more datasource agnostic. Also time is often used for metrics so thought that can be a good thing.

File: public/app/core/logs_model.ts
Comment (marefr) [inline public/app/core/logs_model.ts]: I guess we can skip this check since we're validating that a time and message column exists?

**Review by marefr — COMMENTED** [review]
Some comments on some of the design decision I've made.

File: public/app/core/logs_model.ts
Comment (torkelo) [inline public/app/core/logs_model.ts]: using map on rows is not ideal from a performance perspective, map is 100 times slower than a regular for lop, and there could be 100 to many thousands of rows.

File: public/app/core/logs_model.ts
Comment (torkelo) [inline public/app/core/logs_model.ts]: doing this in the main row loop would be preferable instead of looping in worst case a lot rows again (via callback fn). also why the sortedRows null check? how can it be null?

File: public/app/core/logs_model.ts
Comment (torkelo) [reply public/app/core/logs_model.ts]: For query result that returns many series this seems like it cause 2 array copy per series (as reduce returns a new array for each loop, copying results from the last). Looks a bit inefficient :)

File: public/app/plugins/datasource/loki/datasource.ts
Comment (torkelo) [reply public/app/plugins/datasource/loki/datasource.ts]: think that should be a good use of the meta field

File: public/app/plugins/datasource/loki/result_transformer.ts
Comment (torkelo) [reply public/app/plugins/datasource/loki/result_transformer.ts]: 👍

**Review by torkelo — COMMENTED** [review]
Great work! So nice to see progress on this front!  Just some minor comments about the overuse of callback loops (reduce/maps etc) for arrays that can have thousands of entries.

File: public/app/plugins/datasource/loki/datasource.ts
Comment (ryantxu) [reply public/app/plugins/datasource/loki/datasource.ts]: +1 -- these also seem reasonable to put in the typescript defs rather than just `any`

File: public/app/plugins/datasource/loki/result_transformer.ts
Comment (ryantxu) [reply public/app/plugins/datasource/loki/result_transformer.ts]: time is good -- but it should not matter what name you use since you the type is 'time' and on the other end you should use `getFirstTimeField`

File: public/app/core/logs_model.ts
Comment (ryantxu) [inline public/app/core/logs_model.ts]: Do we need to require that the name is 'message' can it just be the first/longest string?  or something like that?

File: public/app/core/logs_model.ts
Comment (ryantxu) [inline public/app/core/logs_model.ts]: we don't need to solve this now, but it would be great to have the log level possibly come from the query:
`SELECT time,level,messsage FROM ...`

**Comment by ryantxu** [comment]
@marefr -- this looks really good.  thank you.

File: public/app/core/logs_model.ts
Comment (ryantxu) [reply public/app/core/logs_model.ts]: no need for this -- it just makes it harder for other datasources to implement

File: public/app/plugins/datasource/loki/result_transformer.ts
Comment (marefr) [reply public/app/plugins/datasource/loki/result_transformer.ts]: Have fixed this to be name-independent and use time and string field as requirement instead.

File: public/app/core/logs_model.ts
Comment (marefr) [reply public/app/core/logs_model.ts]: Removed

File: public/app/core/logs_model.ts
Comment (marefr) [reply public/app/core/logs_model.ts]: Have tried to fix this. Please review

File: public/app/core/logs_model.ts
Comment (marefr) [reply public/app/core/logs_model.ts]: Have tried to fix this. Please review

File: public/app/core/logs_model.ts
Comment (marefr) [reply public/app/core/logs_model.ts]: Fixed

File: public/app/core/logs_model.ts
Comment (marefr) [reply public/app/core/logs_model.ts]: Have added some logic for this, but probably needs some more love to support other field names than "loglevel".

File: public/app/plugins/datasource/loki/datasource.ts
Comment (marefr) [reply public/app/plugins/datasource/loki/datasource.ts]: Fixed

**Comment by marefr** [comment]
@ryantxu @torkelo thanks for the feedback :+1: Have pushed some changes now. I guess the biggest change is the `SeriesFieldsProcessor` which I did find handy when processing a lot of rows without needing to loop thru all the fields every time to find the correct field indices. 

Besides this have tried to make the log processing more efficient by using regular loops instead.

File: public/app/features/explore/LogsContainer.tsx
Comment (marefr) [inline public/app/features/explore/LogsContainer.tsx]: Removed `id` from `LogsModel` since I don't think the key is needed here - to me should only be needed if you're rendering multiple `<Logs />` components.

File: packages/grafana-ui/src/types/data.ts
Comment (ryantxu) [inline packages/grafana-ui/src/types/data.ts]: Can we move `IndexedField` to `processSeriesData.ts`?  I'm hoping `data.ts` just represents the core data exchange formats and the rest can be more efemeral/helpers

File: packages/grafana-ui/src/utils/processSeriesData.ts
Comment (ryantxu) [inline packages/grafana-ui/src/utils/processSeriesData.ts]: rather than a new array for each type, what about:
```
fieldTypeIndices: { [type: FieldType]: number[] };
```

and than adding FieldType as a parameter to the getters below?  That feels less brittle as we add/refine the various field type options and approaches.

**Comment by ryantxu** [comment]
@marefr this is looking really good -- thanks.

The name on `SeriesFieldsProcessor` feels a bit off, and potentially in conflict with some series processing/transforming work I hope to standardize in #16756

As implemented it is a way to find field indexes based on name/type -- maybe `SeriesFieldLocator` or `SeriesFieldHelper`?

File: public/app/core/logs_model.ts
Comment (hugohaggmark) [inline public/app/core/logs_model.ts]: Maybe we could new up this once and then process the series in the loop?

File: public/app/plugins/datasource/loki/datasource.ts
Comment (hugohaggmark) [inline public/app/plugins/datasource/loki/datasource.ts]: Awesome!

**Review by hugohaggmark — APPROVED** [review]
Awesome work, found some very minor stuff.

File: packages/grafana-ui/src/types/data.ts
Comment (marefr) [reply packages/grafana-ui/src/types/data.ts]: Makes sense :+1: Will fix

File: packages/grafana-ui/src/utils/processSeriesData.ts
Comment (marefr) [reply packages/grafana-ui/src/utils/processSeriesData.ts]: Good point

File: public/app/core/logs_model.ts
Comment (marefr) [reply public/app/core/logs_model.ts]: You mean so that it can be used in line 382 as well, i.e. use an array of processors instead of looping seriesData?

**Comment by marefr** [comment]
> The name on `SeriesFieldsProcessor` feels a bit off, and potentially in conflict with some series processing/transforming work I hope to standardize in #16756
> 
> As implemented it is a way to find field indexes based on name/type -- maybe `FieldIndexFinder` or `IndexFieldFinder`?

@ryantxu I agree, I'm not super happy with that name. But it's not only a way for finding field indexes based on name/type - it also guesses field type for undefined/other field types and "caches" it so you don't have to iterate over all fields over and over again. And for `LogsSeriesFieldProcessor` it adds some extra field type processing to find log level field => the log level finding seems kind of similar to your proposed matchers.

Would it be possible/a good thing to wrap the SeriesData in some sort of class/structure, similar to `SeriesFieldsProcessor`, mixed with your matchers, extensions and transforms and then use this class/structure internally in Grafana instead of SeriesData? 

I guess it's easier to start with renaming `SeriesFieldsProcessor` and get this merged and later if we merge your proposed PR and potential additional matchers/transforms we may be able to easier spot what is missing/how to properly refactor things.

**Comment by marefr** [comment]
@ryantxu I think changes in https://github.com/grafana/grafana/pull/16793/files#diff-317237f74791e054ee61a4e01ed3cb3bR58 and https://github.com/grafana/grafana/pull/16793/files#diff-c10a219f8d98d5afcc02054eb27f23a3R36 highlights the need of using some abstraction on top of `SeriesData`. What do you think?

File: public/app/core/logs_model.ts
Comment (hugohaggmark) [reply public/app/core/logs_model.ts]: Well I was thinking of new:ing up objects in a loop, don't know how many objects there are in this case. As the rest of the discussion seemed to be leaned towards performance I added this. No biggie.

**Comment by ryantxu** [comment]
@marefr -- maybe, I don't think we need more now. regardless, I the details of this class should not hold up merging the PR.

**Comment by ryantxu** [comment]
>  it also guesses field type for undefined/other field types and "caches" it so you don't have to iterate over all fields over and over again.

FYI - this step already happens if you query with the new [PanelQueryRunner](https://github.com/grafana/grafana/blob/master/public/app/features/dashboard/state/PanelQueryState.ts#L347) right now that is only react panels, but soon will be everything.  Essentially it makes sure there is a type on everything before sending it to any panel.

File: public/app/core/logs_model.ts
Comment (marefr) [reply public/app/core/logs_model.ts]: Maximum 28 or something I suppose - the maximum of `refId`, a b c etc.

**Comment by marefr** [comment]
> > it also guesses field type for undefined/other field types and "caches" it so you don't have to iterate over all fields over and over again.
> 
> FYI - this step already happens if you query with the new [PanelQueryRunner](https://github.com/grafana/grafana/blob/master/public/app/features/dashboard/state/PanelQueryState.ts#L347) right now that is only react panels, but soon will be everything. Essentially it makes sure there is a type on everything before sending it to any panel.

I totally see the win and simplicity of using this. Next step for me in explore is to refractor how queries are executed and then I think it will make sense to use the same, `guessFieldTypes(toSeriesData(...))`, and later on add some additional field processing specific for logs. Then the `SeriesDataProcessor` may not be needed or it will be just a `FieldIndexFinder` or whatever we decide to call it. I'll give this a try tomorrow before merging.

**Comment by ryantxu** [comment]
If you get ambitions and subscribe to the results, you may get streaming to work for "free" -- see:
https://github.com/grafana/grafana/pull/16815

File: public/app/core/logs_model.ts
Comment (marefr) [reply public/app/core/logs_model.ts]: This is now gone :boom:

**Comment by ryantxu** [comment]
@marefr -- looks great.  this is a huge improvement.

**Comment by marefr** [comment]
Thanks for great feedback @ryantxu :+1: Really love how the new SeriesData format simplifies a lot of stuff! Next up is #16823 and next after that simplifying query execution in explore so that SeriesData is used for all both metrics and logs.

**Comment by davkal** [comment]
@marefr using indexes as React keys is rarely advisable. The key tells react what to unmount, so now when you look at a loki stream (seeing all results) and then decide to add a search term that would match some of the later results (e.g., a term from row 10) and then run the query, the non-matched results still show. This happens because in the filtered set the first indexes are still present, so react thinks it won't have to unmount those rows.

**Comment by davkal** [comment]
Opened issue: #16832

**Pull request merged**
